### PR TITLE
boot.initrd: add verbose option

### LINKED ISSUE #32555: add option for silent boot
## Issue description

I don't want to see the boot messages on my desktop systems. The Login manager should appear directly after the Grub menu (when there are no critical errors).

I tried several things like:

```
  # less verbose boot log
  boot.consoleLogLevel = 3;
  boot.kernelParams = [ "quiet" "udev.log_priority=3" ];
```
and
```
  boot.loader.grub.extraConfig = "GRUB_CMDLINE_LINUX_DEFAULT=\"quiet loglevel=3 vga=current\"";
```
It has no effect.

Source: https://wiki.archlinux.org/index.php/silent_boot

### Steps to reproduce

Have NixOS system with Grub 2.

virtualbox_nixos_11_12_2017_06_02_11

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.66, NixOS, 17.09.2281.b4a0c011e81 (Hummingbird)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 1.11.15`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

### PR DESCRIPTION
**Pull request by GovanifY**
###### Motivation for this change

This fixes https://github.com/NixOS/nixpkgs/issues/32555

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/system/boot/stage-1.nix [hunk 1]
```diff
@@ -274,6 +274,8 @@ let
       ${pkgs.buildPackages.busybox}/bin/ash -n $target
     '';
 
+    verbose = config.boot.initrd.verbose;
```
Comment (lopsided98) [inline nixos/modules/system/boot/stage-1.nix:277]: Why not add this to the `inherit (config.boot.initrd)` line below?

**Review by lopsided98 — CHANGES_REQUESTED** [review]
Have not tested, but looks good to me except for the one minor suggestion.

File: nixos/modules/system/boot/stage-1.nix (hunk 1 above)
Comment (GovanifY) [reply nixos/modules/system/boot/stage-1.nix]: Because I did not pay a close enough attention to the code style^^" Should be fixed now.

**Comment by GovanifY** [comment]
There seems to be an error on our cachix instance, hence why my test has failed. The failure is meaningless and we can probably merge as is.

**Comment by GovanifY** [comment]
@lopsided98 @oxij @Mic92 Is there any news on the review of this patch?

File: nixos/modules/system/boot/stage-1-init.sh [hunk 2]
```diff
@@ -63,9 +64,11 @@ trap 'fail' 0
 
 
 # Print a greeting.
-echo
-echo "[1;32m<<< NixOS Stage 1 >>>[0m"
-echo
+if [ ! -z "$verbose" ]; then
```
Comment (AndersonTorres) [inline nixos/modules/system/boot/stage-1-init.sh:67]: In general I would suggest to "factor" this pattern:

```shell
if [ -z $verbose ]; then
   echo useful strings
fi
```

It appears three times in this code, and certainly can be useful in other contexts. Maybe something like 

```shell
function printIfVerbose {
  if [ -z $verbose ]; then
    echo $@
  fi
}
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/modules/system/boot/stage-1-init.sh (hunk 2 above)
Comment (Mic92) [reply nixos/modules/system/boot/stage-1-init.sh:67]: I would also have suggested something on those lines:

```diff
diff --git a/nixos/modules/system/boot/stage-1-init.sh b/nixos/modules/system/boot/stage-1-init.sh
index 1578c6cc34c..c80da688707 100644
--- a/nixos/modules/system/boot/stage-1-init.sh
+++ b/nixos/modules/system/boot/stage-1-init.sh
@@ -4,6 +4,12 @@ targetRoot=/mnt-root
 console=tty1
 verbose="@verbose@"
 
+info() {
+    if [[ -n "$verbose" ]]; then
+        echo "$@"
+    fi
+}
+
 extraUtils="@extraUtils@"
 export LD_LIBRARY_PATH=@extraUtils@/lib
 export PATH=@extraUtils@/bin
@@ -56,7 +62,7 @@ EOF
         echo "Rebooting..."
         reboot -f
     else
-        echo "Continuing..."
+        info "Continuing..."
     fi
 }
 
@@ -64,11 +70,9 @@ trap 'fail' 0
 
 
 # Print a greeting.
-if [ ! -z "$verbose" ]; then
-    echo
-    echo "[1;32m<<< NixOS Stage 1 >>>[0m"
-    echo
-fi
+info
+info "[1;32m<<< NixOS Stage 1 >>>[0m"
+info
 
 # Make several required directories.
 mkdir -p /etc/udev
@@ -213,18 +217,14 @@ ln -s @modulesClosure@/lib/modules /lib/modules
 ln -s @modulesClosure@/lib/firmware /lib/firmware
 echo @extraUtils@/bin/modprobe > /proc/sys/kernel/modprobe
 for i in @kernelModules@; do
-    if [ ! -z "$verbose" ]; then
-        echo "loading module $(basename $i)..."
-    fi
+    info "loading module $(basename $i)..."
     modprobe $i
 done
 
 
 # Create device nodes in /dev.
 @preDeviceCommands@
-if [ ! -z "$verbose" ]; then
-    echo "running udev..."
-fi
+info "running udev..."
 ln -sfn /proc/self/fd /dev/fd
 ln -sfn /proc/self/fd/0 /dev/stdin
 ln -sfn /proc/self/fd/1 /dev/stdout
@@ -242,9 +242,7 @@ udevadm settle
 # XXX: Use case usb->lvm will still fail, usb->luks->lvm is covered
 @preLVMCommands@
 
-if [ ! -z "$verbose" ]; then
-    echo "starting device mapper and LVM..."
-fi
+info "starting device mapper and LVM..."
 lvm vgchange -ay
 
 if test -n "$debug1devices"; then fail; fi
```

Comment (GovanifY) [reply nixos/modules/system/boot/stage-1-init.sh]: Thanks to both of you for the review! This should take care of that

**Review by Mic92 — APPROVED** [review]

**Comment by Mic92** [comment]
@GrahamcOfBorg test grub

File: nixos/modules/system/boot/stage-1.nix [hunk 3]
```diff
@@ -560,6 +560,23 @@ in
...
 
+    boot.initrd.verbose = mkOption {
+      default = true;
+      type = types.bool;
+      description =
+        ''
+          Verbosity of the initrd. Please note that disabling verbosity only
+          removes the mandatory messages generated by our scripts. For a fullly
+          silent boot you might want to set the two following configuration
+          options additionally:
```
Comment (AndersonTorres) [inline nixos/modules/system/boot/stage-1.nix:571]:
```suggestion
          Verbosity of the initrd. Please note that disabling verbosity removes 
          only the mandatory messages generated by the NixOS scripts. For a 
          completely silent boot, you might also want to set the following 
          configuration options:
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: nixos/modules/system/boot/stage-1.nix (hunk 3 above)
Comment (AndersonTorres) [reply nixos/modules/system/boot/stage-1.nix:571]: Yes, it is just cosmetical :) but it is way better IMHO.

**Comment by SuperSandro2000** [comment]
> @GrahamcOfBorg test grub

aarch64 failed.

**Comment by AndersonTorres** [comment]
Error log:

https://logs.nix.ci/?key=nixos/nixpkgs.108294&attempt_id=0523dd99-4357-487d-a159-8ff066243862

File: nixos/modules/system/boot/stage-1.nix (hunk 3 above)
Comment (GovanifY) [reply nixos/modules/system/boot/stage-1.nix]: commited, thanks!

**Comment by GovanifY** [comment]
I don't think the aarch64 failure is my fault in this case?

**Comment by AndersonTorres** [comment]
Indeed the error log says something about not being able to find the hard disk (?).

**Comment by AndersonTorres** [comment]
Now LGTBorg.

**Review by AndersonTorres — APPROVED** [review]

**Comment by kanashimia** [comment]
Can't we just redirect output to /dev/kmsg ?
Like loglevels already exist, so why there should be a separate flag for this?

**Comment by GovanifY** [comment]
@kanashimia If you have a better implementation idea go for it, but the patch has already been aproved and is ready for merge

**Comment by AndersonTorres** [comment]
@kanashimia I will merge this PR. Please make a new PR with your suggestions.

**Comment by peterhoeg** [comment]
This is great, thank you! As for @kanashimia's comment, another option would be to read the `quiet` flag off `/proc/cmdline` and act based on that as it allows you to dynamically switch things on/off.

**Pull request merged**